guidance/models/transformers/_transformers.py

```diff
@@ -316,18 +356,38 @@ def get_logits(self, token_ids, forced_bytes, current_temp):
         new_token_ids = token_ids[past_length:]
         if len(new_token_ids) > 0:
             with torch.no_grad():
-                model_out = self.model_obj(
-                    input_ids=torch.tensor(new_token_ids).unsqueeze(0).to(self.device),
-                    past_key_values=self._past_key_values,
-                    use_cache=True,
-                    position_ids=torch.arange(past_length, past_length + len(new_token_ids))
-                    .unsqueeze(0)
-                    .to(self.device),
-                    attention_mask=torch.ones(1, past_length + len(new_token_ids)).to(self.device),
-                    return_dict=True,
-                    output_attentions=False,
-                    output_hidden_states=False,
-                )
+                # Not all models support batched tokens for some reason
+                try:
+                    model_out = self.model_obj(
+                        input_ids=torch.tensor(new_token_ids).unsqueeze(0).to(self.device),
+                        past_key_values=self._past_key_values,
+                        use_cache=True,
+                        position_ids=torch.arange(past_length, past_length + len(new_token_ids))
+                        .unsqueeze(0)
+                        .to(self.device),
+                        attention_mask=torch.ones(1, past_length + len(new_token_ids)).to(self.device),
+                        return_dict=True,
+                        output_attentions=False,
+                        output_hidden_states=False,
+                    )
+                except AssertionError:
+                    for i, new_token_id in enumerate(new_token_ids):
+                        input_ids = torch.tensor([new_token_id]).unsqueeze(0).to(self.device)
+
+                        model_out = self.model_obj(
+                            input_ids=input_ids,
+                            past_key_values=self._past_key_values,
+                            use_cache=True,
+                            position_ids=torch.arange(past_length, past_length + 1).unsqueeze(0).to(self.device),
+                            attention_mask=torch.ones(1, past_length + 1).to(self.device),
+                            return_dict=True,
+                            output_attentions=False,
+                            output_hidden_states=False,
+                        )
+
+                        self._past_key_values = model_out.past_key_values
+                        past_length += 1
+
```

Comment:
This causes a bug with lm.engine.metrics.engine_output_tokens where the tokens being fed one at a time means it's adding together input and output tokens -- the model generates at most 1 token at each gen in the smoke test but the test fails because the output_tokens count is incorrect

```

@guidance
def smoke_test(lm):
    lm.engine.reset_metrics()
    print(lm.engine.metrics.engine_output_tokens)
    lm += "abcd"
    print(lm.engine.metrics.engine_output_tokens)
    lm += gen("first", max_tokens=1)
    print(lm.engine.metrics.engine_output_tokens)
    lm += "fg"
    print(lm.engine.metrics.engine_output_tokens)
    lm += gen("second", max_tokens=1)
    print(lm.engine.metrics.engine_output_tokens)
    return lm
```

With phi-2 + print statements added in _model.Engine.__call__

```
0
0
False // ([50256, 39305], b'd', 0.0) // None
False // None // (b'ef', True, 1.0, {}, {}, 0)
True // None // (b'', False, 1.0, {'first': b'ef'}, {'first': 0.0}, 1)
1
1
False // ([50256, 39305, 2934, 487, 70], b'', 0.0) // None
False // None // (b'gh', True, 1.0, {}, {}, 0)
True // None // (b'', False, 1.0, {'second': b'gh'}, {'second': 0.0}, 1)
2
abcdeffggh
```

with phi-3-small-8k-instruct 

```
0
0
False // ([100257], b'abcd', 0.0) // None
False // ([100257, 64], b'bcd', 0.0) // None
False // ([100257, 64, 65], b'cd', 0.0) // None
True // None // (b'', False, 1.0, {'first': b''}, {'first': 0.0}, 0)
3
3
False // ([100257, 69744], b'fg', 0.0) // None
False // ([100257, 69744, 69], b'g', 0.0) // None
False // None // (b'h', True, 1.0, {}, {}, 0)
True // None // (b'', False, 1.0, {'second': b'h'}, {'second': 0.0}, 1)
5
abcdfgh
```